proposals/lip-0039.md

```diff
@@ -302,273 +328,379 @@ proof={siblingHashes:[cc956a85, 3c516152, e041e1c0, 6400721e, 3c32b131],
-        let h=length of q.binaryBitmap
-        let binaryKey=binaryExpansion(q.key)
-        let q.keyPrefix be the first h digits of binaryKey
-
-        if there exists another p in queries s.t. p.keyPrefix == q.keyPrefix:
-            if p.hash != q.hash:
-                // the two merging queries have mismatching hashes
-                throw error
-            // filter queries by keeping only unique values of q.keyPrefix
-            remove p from queries
+```python
+def isInclusionProof(queryKey: bytes, query: Query) -> bool:
+    return queryKey == query.key and query.value != EMPTY_VALUE
+```
 
-    return queries
+```python
+def smtVerifyInclusionProof(queryKeys: list[bytes], proof: Proof, root: bytes) -> bool:
+    # Check that there is a query for each key.
+    if len(queryKeys) != len(proof.queries):
+        return False
+
+    for key, query in zip(queryKeys, proof.queries):
+        # This is the only difference between an inclusion and a non-inclusion proof.
+        if isInclusionProof(key, query) == False:
+            return False
+
+    return verify(queryKeys, proof, root)
 ```
 
-```java
-// This function checks whether two queries correspond to nodes that are
-// children of the same branch node, with q1 and q2 the left and right
-// child respectively
-function areSiblingQueries(q1, q2):
-    if length of q1.binaryBitmap != length of q2.binaryBitmap:
-        return false
-    let h=length of q1.binaryBitmap
-    let binaryKey1=binaryExpansion(q1.key)
-    let binaryKey2=binaryExpansion(q2.key)
-    let keyPrefix1 be the first (h-1) digits of binaryKey1
-    let keyPrefix2 be the first (h-1) digits of binaryKey2
-    if keyPrefix1 != keyPrefix2:
-        return false
-    let d1 be the digit at position h of binaryKey1
-    let d2 be the digit at position h of binaryKey2
-    return d1==0 and d2==1
+```python
+def smtVerifyNonInclusionProof(queryKeys: list[bytes], proof: Proof, root: bytes) -> bool:
+    # Check that there is a query for each key.
+    if len(queryKeys) != len(proof.queries):
+        return False
+
+    for key, query in zip(queryKeys, proof.queries):
+        # This is the only difference between an inclusion and a non-inclusion proof.
+        if isInclusionProof(key, query) == True:
+            return False
+            
+    return verify(queryKeys, proof, root)
 ```
 
-```java
-function calculateRoot(siblingHashes, queries):
-    // we add the hash property to each query, used to recompute the tree root
-    for each query q in queries:
-        set q.binaryBitmap to binaryExpansion(q.bitmap) without leading zeros
-        // if value is empty, it is a defaultEmptyNode
-        if q.value is empty bytes:
-            set q.hash to EMPTY_HASH
+```python
+def verify(queryKeys: list[bytes], proof: Proof, root: bytes) -> bool:
+    for key, query in zip(queryKeys, proof.queries):
+        # The bitmap does not contain extra leading 0s.
+        if query.bitmap[0] == 0:
+            return False
```

Comment:
This means that verify on a tree containing just one key-value pair will fail.

Reply:
in that case I think the proof queries should be empty, as the key value would be the root

Reply:
Assume, the STM contains only the key value pari `(a,b)`. I then call `generateProof(a, [a]) and obtain
```
Proof:
    siblingHashes = []
    queries = [Query( key = a, value =b, bitmap = 0x00]
```

Reply:
This is maybe not well defined then. I would imagine that the bitmap should be empty in this case. I will test this again

Reply:
I get an empty bitmap in the case you describe (and indeed I added an extra check to fix the edge case)

Reply:
Okay, works now.